spec/forms/applicants/previous_reference_form_spec.rb

```diff
@@ -0,0 +1,150 @@
+      end
+    end
+
+    context "with yes chosen but invalid previous reference number provided" do
+      context "with too many numbers" do
+        let(:params) { { model: applicant, applied_previously: "true", previous_reference: "3234567890123" } }
+
+        it "is invalid" do
+          call_save
+          expect(instance).not_to be_valid
+        end
+
+        it "adds custom blank error message" do
+          call_save
+          error_messages = instance.errors.messages.values.flatten
+          expect(error_messages).to include("Enter the reference number in the correct format")
+        end
+      end
+
+      context "with not enough numbers" do
+        let(:params) { { model: applicant, applied_previously: "true", previous_reference: "3234567890" } }
+
+        it "is invalid" do
+          call_save
+          expect(instance).not_to be_valid
+        end
+
+        it "adds custom blank error message" do
+          call_save
+          error_messages = instance.errors.messages.values.flatten
+          expect(error_messages).to include("Enter the reference number in the correct format")
+        end
+      end
+
+      context "when a non-integer included" do # rubocop:disable RSpec/RepeatedExampleGroupBody
+        let(:params) { { model: applicant, applied_previously: "true", previous_reference: "32345678901F" } }
+
+        it "is invalid" do
+          call_save
+          expect(instance).not_to be_valid
+        end
+
+        it "adds custom blank error message" do
+          call_save
+          error_messages = instance.errors.messages.values.flatten
+          expect(error_messages).to include("Enter the reference number in the correct format")
+        end
+      end
+
+      context "when 1 or 3 is not at the beginning" do # rubocop:disable RSpec/RepeatedExampleGroupBody
+        let(:params) { { model: applicant, applied_previously: "true", previous_reference: "32345678901F" } }
+
+        it "is invalid" do
+          call_save
+          expect(instance).not_to be_valid
+        end
+
+        it "adds custom blank error message" do
+          call_save
+          error_messages = instance.errors.messages.values.flatten
+          expect(error_messages).to include("Enter the reference number in the correct format")
+        end
+      end
+    end
+  end
+
+  describe "#save_as_draft" do
+    subject(:save_as_draft) { instance.save_as_draft }
+
+    context "when yes is chosen but no previous_reference is provided" do
+      let(:params) { { model: applicant, applied_previously: "true", previous_reference: "" } }
+
+      it "is valid" do
+        save_as_draft
+        expect(instance).to be_valid
+      end
+
+      it "updates the applied_previously field in the database" do
+        expect { save_as_draft }.to change { applicant.attributes.symbolize_keys }
+          .from(hash_including(applied_previously: nil))
+          .to(hash_including(applied_previously: true))
+      end
+    end
```

Comment:
I think we could do with another test for save as draft, if the user doesn't select yes or no, something like this

```
context "when neither yes nor no is chosen for applied_previously question" do
      let(:params) { { model: applicant, applied_previously: "", previous_reference: "" } }

      it "is valid" do
        save_as_draft
        expect(instance).to be_valid
      end
   end
```

Reply:
Thank you for the suggestion. I have added this test now.